rfcs/007-background-workers.md

```diff
@@ -0,0 +1,299 @@
+contention during insertion into the queue.
+
+Workers monitor the pending queue of their matching `JobType`. That is,
+`"replication"` workers will monitor the `"replication"` pending queue,
+indexing workers the `"indexing"` queue, etc. If there are jobs ready to run,
+they will attempt to grab one or more jobs from the pending queue and assign it
+to themselves for execution. They do that by setting the `WorkerId` in the jobs
+table entry to their worker ID, removing the job from pending queue and creating
+a new entry in their own `"active"` jobs area.
+
+When a job finishes running, either because it was a successful completion or a
+terminal failure, the worker MUST remove the job from its active table and
+clear its `WorkerId` field. Then, based on type-specific behavior, it MUST do
+either one of:
+
+ - Update the `JobState` to `"completed"` and leave the job in the jobs table.
+   It would be up to the job creation logic to inspect and remove it.
+
+ - Delete the job from the system altogether. This may be useful when running
+   `_replicate` jobs, for example.
+
+
+If a user decided to cancel a job that is running, they MUST toggle `CancelReq`
+job field to `true`. Then, they MUST wait for the job to stop running. If the
+job is not running, they may directly remove the job from the pending and the
+main jobs table.
+
+### Worker Capacity and Long Running Jobs
+
+Some background jobs, like continuous replication jobs, might run for years
+without interruption once they start executing. In that case, workers MAY
+choose to periodically stop some jobs and re-insert them into the pending queue
+for execution, and choose to run other jobs which have been waiting. This
+behavior is job type specific. Ensuring fairness and preventing starvation is
+also job type-specific.
+
+Worker MAY decide how often and how many jobs they pull from the pending queue.
+For example, they could use a dampening scheme where the longer their active
+queue grows, the slower the rate at which they pull jobs from the queue. That
+behavior is job type specific.
+
+Another advanced scheme is where workers could start on demand based on
+the pending queue backlog. When the backlog subsides they would stop
+(with appropriate dampening factors to reduce flapping). That is again job type
+specific, and is outside the scope of this RFC.
+
+## Worker Lifecyle
+
+Workers can be dynamically started and stopped. When workers start, they MUST
+register themselves in the `"workers"` subspace based on their `WorkerType`.
+Each change to the workers list, should be accompanied by updating the
+`"workers_vs"` versionstamp. That versionstamp MAY be efficiently monitored for
+changes using a watch.
+
+
+### Worker Liveness Monitoring
+
+As long as the workers are active and ready to process jobs, they MUST
+periodically ping their health key. The period MUST be less than the
+`HealthTimeout` value.
+
+Worker nodes MUST monitor their neighbors for liveness. When they detect an
```

Comment:
this seems odd to me. Given the central registration, shouldn't the job of monitoring be done generically by some agent(s)? The act of updating a key to indicate liveness and another property indicating how out of date it can get seemed to indicate that monitoring was part of the framework and not the responsibility of everyone that uses the framework.

Also 'neighbour' was not defined before its first mention.

Reply:
Good point. I had flip-flopped in the previous version whether this should be job type specific (replication jobs know how to clean stale replication workers' jobs, indexers indexing jobs etc). But I think it can be done by the framework. So will update the wording there.

Also neighbour should be defined as a term perhaps. Will do that.

Reply:
I update the RFC to have a framework API where it is hopefully more clear where the monitoring happens.